src/txn/txn_global.c

```diff
@@ -0,0 +1,627 @@
+/*
+ * __wt_txn_publish_metadata_pinned --
+ *	Publish a transaction's metadata pinned transaction ID.
+ */
+void
+__wt_txn_publish_metadata_pinned(WT_SESSION_IMPL *session)
+{
+	WT_TXN *prev, *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	WT_ASSERT(session, txn->metadata_pinned != WT_TXN_NONE &&
+	    !F_ISSET(txn, WT_TXN_PUBLIC_METADATA_PINNED));
+
+	__wt_writelock(session, &txn_global->metadata_pinned_rwlock);
+	for (prev = TAILQ_LAST(&txn_global->metadata_pinnedh, __wt_txn_id_qh);
+	    prev != NULL &&
+	    WT_TXNID_LT(txn->metadata_pinned, prev->metadata_pinned);
+	    prev = TAILQ_PREV(prev, __wt_txn_mp_qh, metadata_pinnedq))
+		;
+	if (prev == NULL)
+		TAILQ_INSERT_HEAD(
+		    &txn_global->metadata_pinnedh, txn, metadata_pinnedq);
+	else
+		TAILQ_INSERT_AFTER(
+		    &txn_global->metadata_pinnedh, prev, txn, metadata_pinnedq);
+	__wt_writeunlock(session, &txn_global->metadata_pinned_rwlock);
+	F_SET(txn, WT_TXN_PUBLIC_METADATA_PINNED);
+}
+
+/*
+ * __wt_txn_clear_metadata_pinned --
+ *	Clear a transaction's published pinned ID.
+ */
+void
+__wt_txn_clear_metadata_pinned(WT_SESSION_IMPL *session)
+{
+	WT_TXN *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	if (!F_ISSET(txn, WT_TXN_PUBLIC_METADATA_PINNED))
+		return;
+
+	__wt_writelock(session, &txn_global->metadata_pinned_rwlock);
+	TAILQ_REMOVE(&txn_global->metadata_pinnedh, txn, metadata_pinnedq);
+	txn->metadata_pinned = WT_TXN_NONE;
+	__wt_writeunlock(session, &txn_global->metadata_pinned_rwlock);
+	F_CLR(txn, WT_TXN_PUBLIC_METADATA_PINNED);
+}
+
+/*
+ * __wt_txn_publish_pinned_id --
+ *	Publish a transaction's pinned transaction ID.
+ */
+void
+__wt_txn_publish_pinned_id(WT_SESSION_IMPL *session)
+{
+	WT_TXN *prev, *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	WT_ASSERT(session, txn->pinned_id != WT_TXN_NONE &&
+	    !F_ISSET(txn, WT_TXN_PUBLIC_PINNED_ID));
+
+	__wt_writelock(session, &txn_global->pinned_id_rwlock);
+	for (prev = TAILQ_LAST(&txn_global->pinned_idh, __wt_txn_id_qh);
+	    prev != NULL && WT_TXNID_LT(txn->pinned_id, prev->pinned_id);
+	    prev = TAILQ_PREV(prev, __wt_txn_cts_qh, pinned_idq))
+		;
+	if (prev == NULL)
+		TAILQ_INSERT_HEAD(&txn_global->pinned_idh, txn, pinned_idq);
+	else
+		TAILQ_INSERT_AFTER(
+		    &txn_global->pinned_idh, prev, txn, pinned_idq);
+	__wt_writeunlock(session, &txn_global->pinned_id_rwlock);
+	F_SET(txn, WT_TXN_PUBLIC_PINNED_ID);
+}
+
+/*
+ * __wt_txn_clear_pinned_id --
+ *	Clear a transaction's published pinned ID.
+ */
+void
+__wt_txn_clear_pinned_id(WT_SESSION_IMPL *session)
+{
+	WT_TXN *txn;
+	WT_TXN_GLOBAL *txn_global;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	if (!F_ISSET(txn, WT_TXN_PUBLIC_PINNED_ID))
+		return;
+
+	__wt_writelock(session, &txn_global->pinned_id_rwlock);
+	TAILQ_REMOVE(&txn_global->pinned_idh, txn, pinned_idq);
```

Comment:
Since the F_CLR happens outside the lock - should this re-check that there is an ID pinned before removing? Or does TAILQ_REMOVE handle the case where the item doesn't exist?